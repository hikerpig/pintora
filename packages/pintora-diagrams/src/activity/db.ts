import { makeIdCounter } from '@pintora/core'
import { BaseDb } from '../util/base-db'
import { BaseDiagramIR } from '../util/ir'
import { OverrideConfigAction, ParamAction, SetTitleAction } from '../util/config'
import { dedent } from '../util/text'

export type Action = {
  id: string
  actionType: string
  message: string
}

export type Condition = {
  id: string
  message: string
  then: {
    label: string
    children: Step[]
  }
  else?: {
    label: string
    children: Step[]
  }
}

export type While = {
  id: string
  message: string
  children: Step[]
  confirmLabel: string | undefined
  denyLabel: string | undefined
}

export type Repeat = {
  id: string
  message: string
  confirmLabel: string | undefined
  firstAction: Action | undefined
  denyLabel: string | undefined
  children: Step[]
}

export type Keyword = {
  id: string
  label: string
}

/** element group */
export type AGroup = {
  id: string
  groupType: string
  name: string
  label?: string
  children: Step[]
  parent?: string
  background: string | null
}

export type Switch = {
  id: string
  message: string | undefined
  children: Step[]
}

export type Case = {
  id: string
  confirmLabel: string | undefined
  children: Step[]
}

export type Fork = {
  id: string
  branches: Step[]
  shouldMerge: boolean
}

export type ForkBranch = {
  id: string
  children: Step[]
}
export type Note = {
  id: string
  text: string
  placement: string
  target?: string
}

export type ArrowLabel = {
  id: string
  text: string
  target?: string
}

type StepValue = Action | Condition | While | Repeat | Keyword | AGroup | Switch | Case | Fork | ForkBranch | ArrowLabel

export type Step<T extends StepValue = StepValue> = {
  type: string
  parentId?: string
  value: T
}

export type ActivityDiagramIR = BaseDiagramIR & {
  steps: Step[]
  notes: Note[]
  arrowLabels: ArrowLabel[]
}

export type ApplyPart =
  | ParamAction
  | OverrideConfigAction
  | {
      type: 'addAction'
      action: Action
    }
  | {
      type: 'condition'
      message: string
      then: {
        label: string
        children: ApplyPart[]
      }
      else?: {
        label: string
        children: ApplyPart[]
      }
    }
  | {
      type: 'while'
      message: string
      confirmLabel: string
      denyLabel: string
      children: ApplyPart[]
    }
  | {
      type: 'repeat'
      message: string
      confirmLabel: string
      denyLabel: string
      firstAction: Action
      children: ApplyPart[]
    }
  | {
      type: 'switch'
      message: string
      children: ApplyPart[]
    }
  | {
      type: 'case'
      confirmLabel: string
      children: ApplyPart[]
    }
  | {
      type: 'keyword'
      label: string
    }
  | {
      type: 'group'
      groupType: string
      name: string
      label?: string
      children: ApplyPart[]
      background: string | null
    }
  | {
      type: 'note'
      text: string
      placement: string
    }
  | {
      type: 'arrowLabel'
      text: string
    }
  | {
      type: 'fork'
      branches: ApplyPart[]
      shouldMerge: boolean
    }
  | {
      type: 'forkBranch'
      children: ApplyPart[]
    }
  | SetTitleAction

type DbApplyState = {
  prevStepId?: string | undefined
  parentId?: string | undefined
}

class ActivityDb extends BaseDb {
  protected steps: Step[] = []
  protected notes: Note[] = []
  protected arrowLabels: ArrowLabel[] = []
  protected idCounter = makeIdCounter()

  protected makeId() {
    return this.idCounter.next()
  }

  apply(part: ApplyPart, ignoreAdd?: boolean, state?: DbApplyState): Step
  apply(part: ApplyPart[], ignoreAdd?: boolean, state?: DbApplyState): Step[]
  apply(part: ApplyPart | ApplyPart[], ignoreAdd = false, state: DbApplyState = {}): Step | Step[] {
    if (Array.isArray(part)) {
      const partResults: any[] = []
      let currentStep: Step
      part.forEach(p => {
        const result = this.apply(p, ignoreAdd, { ...state, prevStepId: currentStep?.value?.id })
        if (result) {
          partResults.push(result)
          currentStep = result
        }
      })
      return partResults
    }
    if (!part) return
    // console.log('[activity] apply', part)
    let step: Step
    switch (part.type) {
      case 'addAction':
        {
          const action = { ...part.action, id: this.makeId() }
          step = { type: 'action', value: action }
        }
        break
      case 'condition':
        {
          const id = this.makeId()
          const thenResult = this.apply(part.then.children, true, { ...state, parentId: id })
          const condition: Condition = {
            id,
            message: part.message,
            then: {
              label: part.then.label,
              children: thenResult,
            },
          }
          if (part.else) {
            const elseResult = this.apply(part.else.children, true, { ...state, parentId: id })
            condition.else = {
              label: part.else.label,
              children: elseResult,
            }
          }
          step = { type: 'condition', value: condition }
        }
        break
      case 'while': {
        const id = this.makeId()
        const loopResult = this.apply(part.children, true, { ...state, parentId: id })
        const whileSentence: While = {
          id,
          message: part.message,
          children: loopResult,
          confirmLabel: part.confirmLabel,
          denyLabel: part.denyLabel,
        }
        step = { type: 'while', value: whileSentence }
        break
      }
      case 'repeat': {
        const id = this.makeId()
        const loopResult = this.apply(part.children, true, { ...state, parentId: id })
        const sentence: Repeat = {
          id,
          message: part.message,
          children: loopResult,
          confirmLabel: part.confirmLabel,
          denyLabel: part.denyLabel,
          firstAction: part.firstAction,
        }
        step = { type: 'repeat', value: sentence }
        break
      }
      case 'switch': {
        const id = this.makeId()
        const cases = this.apply(part.children, true, { ...state, parentId: id })
        const switchSentence: Switch = {
          id,
          message: part.message,
          children: cases as any,
        }
        step = { type: 'switch', value: switchSentence }
        break
      }
      case 'case': {
        const id = this.makeId()
        const children = this.apply(part.children, true, { ...state, parentId: id })
        const caseClause: Case = {
          id,
          confirmLabel: part.confirmLabel,
          children,
        }
        step = { type: 'case', value: caseClause }
        break
      }
      case 'fork': {
        const id = this.makeId()
        const branches = this.apply(part.branches, true, { ...state, parentId: id })
        const forkSentence: Fork = {
          id,
          shouldMerge: part.shouldMerge,
          branches,
        }
        step = { type: 'fork', value: forkSentence }
        break
      }
      case 'forkBranch': {
        const id = this.makeId()
        const children = this.apply(part.children, true, { ...state, parentId: id })
        const forkBranch: ForkBranch = {
          id,
          children,
        }
        step = { type: 'forkBranch', value: forkBranch }
        break
      }
      case 'keyword': {
        step = { type: 'keyword', value: { id: this.makeId(), label: part.label } }
        break
      }
      case 'group': {
        const id = this.makeId()
        const childrenResult = this.apply(part.children, true, { ...state, parentId: id })
        step = { type: 'group', value: { id, ...part, children: childrenResult } }
        break
      }
      case 'note': {
        const value: Note = { id: this.makeId(), ...part, text: dedent(part.text) }
        const prevStepId = state.prevStepId
        if (prevStepId) {
          value.target = prevStepId
        }
        this.notes.push(value)
        break
      }
      case 'arrowLabel': {
        const value: ArrowLabel = { id: this.makeId(), ...part }
        const prevStepId = state.prevStepId
        if (prevStepId) {
          value.target = prevStepId
        }
        this.arrowLabels.push(value)
        break
      }
      case 'setTitle': {
        this.title = part.text
        break
      }
      case 'addParam': {
        this.configParams.push(part)
        break
      }
      case 'overrideConfig': {
        this.addOverrideConfig(part)
        break
      }
      default: {
      }
    }
    if (step && !ignoreAdd) {
      this.steps.push(step)
    }
    if (step) {
      step.parentId = state.parentId
    }
    return step as any
  }

  getDiagramIR(): ActivityDiagramIR {
    return {
      steps: this.steps,
      notes: this.notes,
      arrowLabels: this.arrowLabels,
      ...this.getBaseDiagramIR(),
    }
  }

  clear() {
    super.clear()
    this.idCounter.reset()
    this.steps = []
    this.notes = []
  }
}

export const db = new ActivityDb()
