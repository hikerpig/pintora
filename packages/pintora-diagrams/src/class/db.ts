/* eslint-disable @typescript-eslint/no-empty-function */
import { BaseDiagramIR } from '../util/ir'
import { ActionHandler, BaseDb, MakeAction } from '../util/base-db'
import { ConfigParam } from '@pintora/core'
import { OverrideConfigAction } from '../util/config'

/** type to represent one class  */
export type TClass = {
  name: string
  fullName: string
  label?: string
  namespace?: string
  members: TClassMember[]
  annotation?: string
}

export type ClassRelation = {
  left: string
  right: string
  relation: Relation
  label?: string
  labelLeft?: string
  labelRight?: string
  dashed?: boolean
  /**
   * normally we draw the relation from left to right, if reversed is true, we draw the relation from right to left
   */
  reversed?: boolean
}

type Access = 'public' | 'private' | 'protected'

type Modifier = 'abstract' | 'static' | null

export type TClassMember = {
  name: string
  typeName: string
  raw: string
  access?: Access
  isMethod?: boolean
  modifier?: Modifier
}

export enum Relation {
  INHERITANCE = 'INHERITANCE',
  COMPOSITION = 'COMPOSITION',
  AGGREGATION = 'AGGREGATION',
  ASSOCIATION = 'ASSOCIATION',
  LINK = 'LINK',
}

type ClassActionPayloads = {
  addClass: Pick<TClass, 'name'> & {
    members: Array<RawMember | RawAnnotation>
    label?: string
  }
  addClassMember: { className: string; member: RawMember }
  addRelation: { left: string; right: string; relationRaw: { type: Relation; dashed: boolean } }
  addAnnotation: { className: string; annotation: string }
  overrideConfig: OverrideConfigAction
  addParam: ConfigParam
}

type RawMember = {
  raw: string
  modifier: Modifier
}

type RawAnnotation = {
  annotation: string
}

export type Action = MakeAction<ClassActionPayloads>

export type ClassIR = BaseDiagramIR & {
  classes: Record<string, TClass>
  relations: ClassRelation[]
}

const NAMESPACE_SEP = '.'
const FIELD_SEP = ':'

export class ClassDb extends BaseDb {
  protected classes: Record<string, TClass> = {}
  protected relations: ClassRelation[] = []

  ACTION_HANDLERS: { [K in keyof ClassActionPayloads]: ActionHandler<ClassActionPayloads, ClassDb, K> } = {
    addClass(action) {
      let classObj: TClass = this.classes[action.name]
      const { type, ...restData } = action
      const data = this.parseClassAction(restData)
      if (!classObj) {
        classObj = data
      } else {
        Object.assign(classObj, data)
      }
      this.classes[action.name] = classObj
    },
    addClassMember(action) {
      const member = this.parseMemberLabel(action.member.raw)
      let classObj = this.classes[action.className]
      if (!classObj) {
        classObj = this.parseClassAction({ name: action.className, members: [] })
        this.classes[action.className] = classObj
      }
      classObj.members.push(member)
    },
    addRelation(action) {
      let leftClass = this.classes[action.left]
      if (!leftClass) {
        leftClass = this.parseClassAction({ name: action.left, members: [] })
        this.classes[action.left] = leftClass
      }

      let rightClass = this.classes[action.right]
      if (!rightClass) {
        rightClass = this.parseClassAction({ name: action.right, members: [] })
        this.classes[action.right] = rightClass
      }
      const { relationRaw, ...other } = action

      const relation: Partial<ClassRelation> = { ...other, relation: relationRaw.type }
      if (relationRaw.dashed) {
        relation.dashed = relationRaw.dashed
      }
      this.relations.push(relation as ClassRelation)
    },
    addAnnotation(action) {
      let classObj = this.classes[action.className]
      if (!classObj) {
        classObj = this.parseClassAction({ name: action.className, members: [] })
        this.classes[action.className] = classObj
      }
      classObj.annotation = action.annotation
    },
    overrideConfig(action) {
      this.addOverrideConfig(action)
    },
    addParam(action) {
      this.configParams.push(action)
    },
  }

  protected parseClassAction(payload: ClassActionPayloads['addClass']): TClass {
    const fullName = payload.name
    let name = payload.name
    let namespace = ''
    let annotation = ''
    const members: TClassMember[] = []
    if (payload.name.includes(NAMESPACE_SEP)) {
      const segs = payload.name.split(NAMESPACE_SEP)
      name = segs[segs.length - 1]
      namespace = segs.slice(0, segs.length - 1).join(NAMESPACE_SEP)
    }
    if (payload.members) {
      for (const m of payload.members) {
        if ('annotation' in m) {
          annotation = m.annotation
        } else {
          const parsedMember = this.parseMemberLabel(m.raw)
          parsedMember.modifier = m.modifier
          members.push(parsedMember)
        }
      }
    }
    const label = payload.label || name
    return {
      ...payload,
      name,
      namespace,
      fullName,
      members,
      annotation,
      label,
    }
  }

  protected parseMemberLabel(raw: string) {
    let name: string
    let typeName = ''
    let temp = raw
    const firstChar = temp[0]
    let access: Access = 'public'
    const isPrivate = firstChar === '-'
    const isProtected = firstChar === '#'
    const isPublic = firstChar === '+'

    if (isPrivate) access = 'private'
    else if (isProtected) access = 'protected'

    if (isPrivate || isProtected || isPublic) temp = temp.slice(1)

    if (temp.includes(FIELD_SEP)) {
      const pos = temp.indexOf(FIELD_SEP)
      name = temp.slice(0, pos)
      typeName = temp.slice(pos + 1, temp.length).trim()
    } else {
      const spacePos = temp.indexOf(' ')
      if (spacePos === -1) {
        name = temp.trim()
      } else {
        typeName = temp.slice(0, spacePos)
        name = temp.slice(spacePos + 1, temp.length).trim()
      }
    }

    const isMethod = /\(.*\)/.test(name)

    const member: TClassMember = {
      name,
      typeName,
      access,
      raw,
      isMethod,
    }
    return member
  }

  getDiagramIR(): ClassIR {
    return {
      ...super.getBaseDiagramIR(),
      classes: this.classes,
      relations: this.relations,
    }
  }

  apply(action: Action | Action[]) {
    if (!action) return
    if (Array.isArray(action)) {
      action.forEach(a => this.apply(a))
      return
    }
    if (action.type in this.ACTION_HANDLERS) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.ACTION_HANDLERS[action.type].call(this, action as any)
    }
  }

  override clear() {
    super.clear()
    this.classes = {}
    this.relations = []
  }
}

const db = new ClassDb()

export default db
