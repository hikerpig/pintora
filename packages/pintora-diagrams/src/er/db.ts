import { BaseDiagramIR } from '../util/ir'
import { ConfigParam } from '../util/config'
import { BaseDb } from '../util/base-db'

export enum Cardinality {
  ZERO_OR_ONE = 'ZERO_OR_ONE',
  ZERO_OR_MORE = 'ZERO_OR_MORE',
  ONE_OR_MORE = 'ONE_OR_MORE',
  ONLY_ONE = 'ONLY_ONE',
  MORE = 'MORE',
}

export enum Identification {
  NON_IDENTIFYING = 'NON_IDENTIFYING',
  IDENTIFYING = 'IDENTIFYING',
}

export type Attribute = {
  attributeType: string
  attributeName: string
  attributeKey?: string
  comment?: string
}

export type Entity = {
  itemId: string
  attributes: Attribute[]
}

export type Relationship = {
  itemId: string
  entityA: string
  roleA: string
  entityB: string
  relSpec: RelSpec
}

export type RelSpec = {
  cardA: Cardinality
  cardB: Cardinality
  relType: Identification
}

export type Inheritance = {
  sup: string
  sub: string
}

export type ErDiagramIR = BaseDiagramIR & {
  entities: Record<string, Entity>
  relationships: Relationship[]
  inheritances: Inheritance[]
}

export class ErDb extends BaseDb {
  Cardinality = Cardinality
  Identification = Identification

  entities: Record<string, Entity> = {}
  relationships: Relationship[] = []
  inheritances: Inheritance[] = []

  addEntity(name: string) {
    if (!this.entities[name]) {
      this.entities[name] = { attributes: [], itemId: `entity-${name}` }
    }
    return this.entities[name]
  }
  addRelationship(entityA: string, roleA: string, entityB: string, relSpec: RelSpec) {
    const itemId = `relationship-${entityA}-${entityB}`
    const rel: Relationship = {
      entityA,
      roleA,
      entityB,
      relSpec,
      itemId,
    }

    this.relationships.push(rel)
  }
  addInheritance(sup: string, sub: string) {
    this.inheritances.push({ sup, sub })
  }
  getDiagramIR(): ErDiagramIR {
    return {
      ...super.getBaseDiagramIR(),
      entities: this.entities,
      relationships: this.relationships,
      inheritances: this.inheritances,
    }
  }
  addTitle(title: string) {
    this.title = title
  }
  addAttributes(name: string, attributes: Attribute[]) {
    const entity = this.addEntity(name)
    entity.attributes.push(...attributes)
  }
  addParam(styleParam: ConfigParam) {
    this.configParams.push(styleParam)
  }

  override clear() {
    super.clear()
    this.entities = {}
    this.relationships = []
    this.inheritances = []
    this.configParams = []
    this.title = ''
  }
}

const db = new ErDb()

export default db
