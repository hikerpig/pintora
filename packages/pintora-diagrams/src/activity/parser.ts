import { db } from './db'
import grammar from './parser/activityDiagram'
import { genParserWithRules } from '../util/parser-util'

export const parse = genParserWithRules(grammar, {
  dedupeAmbigousResults: true,
  postProcess(results) {
    db.apply(results)
    return results
  },
})
