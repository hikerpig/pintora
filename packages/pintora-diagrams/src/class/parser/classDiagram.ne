@preprocessor typescript
@lexer lexer
@skip_unmatch %WS
@include "whitespace.ne"
@include "config.ne"
@include "comment.ne"

@{%
import * as moo from '@hikerpig/moo'
import {
  tv,
  textToCaseInsensitiveRegex,
  VALID_TEXT_REGEXP,
  COMMENT_LINE_REGEXP,
  QUOTED_WORD_REGEXP,
  configLexerMainState,
  configLexerconfigStatementState,
  L_PAREN_REGEXP,
  R_PAREN_REGEXP,
  COLOR_REGEXP,
  MOO_NEWLINE,
  getQuotedWord,
  makeNth,
} from '../../util/parser-shared'
import type { Action } from '../db'
import { Relation } from '../db'

const COMMON_TOKEN_RULES = {
  VALID_TEXT: { match: VALID_TEXT_REGEXP, fallback: true },
}

let lexer = moo.states({
  main: {
    NL: MOO_NEWLINE,
    WS: { match: /[ \t]+/, lineBreaks: false },
    QUOTED_WORD: QUOTED_WORD_REGEXP,
    COLOR: COLOR_REGEXP,
    SEMICOLON: /;/,
    COLON: /:/,
    COMMA: /,/,
    CLASS_DIAGRAM: /classDiagram/,
    L_PAREN: L_PAREN_REGEXP,
    R_PAREN: R_PAREN_REGEXP,
    L_BRACKET: { match: /\{/ },
    R_BRACKET: { match: /\}/ },
    L_SQ_BRACKET: { match: /\[/ },
    R_SQ_BRACKET: { match: /\]/ },
    TEXT_WITH_ANGLE_BRACKETS: { match: /\<\<(?:.*)\>\>/ },
    EQ: { match: /=/ },
    // RELATION_INHERITANCE: { match: /\<\|\-\-/ },
    SUBGRAPH: { match: /subgraph/ },
    START_NOTE: textToCaseInsensitiveRegex('@note'),
    END_NOTE: textToCaseInsensitiveRegex('@end_note'),
    COMMENT_LINE: COMMENT_LINE_REGEXP,
    ...configLexerMainState,
    VALID_TEXT: { match: VALID_TEXT_REGEXP, fallback: true },
  },
  configStatement: {
    ...configLexerconfigStatementState,
    ...COMMON_TOKEN_RULES,
  },
})

function rNull() {
  return null
}

const nth0 = makeNth(0)
const nth1 = makeNth(1)
%}

start -> __ start {% nth1 %}
	| %CLASS_DIAGRAM document {%
      function(d) {
        return d[1]
      }
    %}

document -> null
  | document statementWrap {%
    (d) => {
        let r = d[0]
        if (d[1]) {
          r = d[0].concat(d[1])
        }
        return r
      }
    %}

statementWrap ->
	  %WS:* statement {% (d) => {
      return d[1]
    } %}
	| %WS:? %NL {% null %}

statement ->
    classStatement
  | memberLabelStatement
  | relationStatement
  | classAnnotationStatement
  | paramStatement %NL
  | configOpenCloseStatement %NL
  | comment %NL

classStatement ->
    "class" %VALID_TEXT %L_BRACKET %NL:? classMembers:? %NL:? %R_BRACKET %NL {%
      function(d) {
        const members = d[4]
        return {
          type: 'addClass',
          name: tv(d[1]),
          members,
        } as Action
      }
    %}
  | "class" textInQuote "as" %VALID_TEXT %NL {%
      function(d) {
        return {
          type: 'addClass',
          name: tv(d[3]),
          label: d[1],
          members: [],
        } as Action
      }
    %}
  | "class" %VALID_TEXT %NL {%
      function(d) {
        return {
          type: 'addClass',
          name: tv(d[1]),
          members: [],
        } as Action
      }
    %}

classMembers ->
    classMember
  | annotation
  | classMembers %NL (annotation | classMember) {%
      function(d) {
        return [...d[0], ...d[2]]
      }
    %}

classMember ->
    modifier:? memberLabel {%
      function(d) {
        const modifier = d[0] ? d[0] : null
        return {
          raw: d[1],
          modifier,
        }
      }
    %}

words ->
    %VALID_TEXT (%VALID_TEXT | %WS):* {%
      function(d) {
        return tv(d[0]) + d[1].map(o => tv(o[0])).join('')
      }
    %}

memberLabel ->
    %VALID_TEXT (%VALID_TEXT | %L_PAREN | %R_PAREN | %COLON | %WS):* {%
      function(d) {
        return tv(d[0]) + d[1].map(o => tv(o[0])).join('')
      }
    %}

modifier -> %L_BRACKET ("static" | "abstract") %R_BRACKET {%
      function(d) {
        return tv(d[1][0])
      }
    %}

memberLabelStatement ->
    %VALID_TEXT %COLON classMember %NL {%
        function(d) {
          const className = tv(d[0])
          return {
            type: 'addClassMember',
            className,
            member: d[2]
          } as Action
        }
      %}

relationStatement ->
    classInRelation textInQuote:? relation textInQuote:? classInRelation (%WS:* %COLON %WS:* words):? {%
        function(d) {
          let relationRaw = { type: d[2], dashed: false, reversed: false }
          let labelLeft = d[1]
          let labelRight = d[3]
          if (d[2].type) {
            relationRaw = d[2]
          }
          let label = ''
          if (d[5]) {
            label = d[5][3]
          }
          return {
            type: 'addRelation',
            left: d[0].name,
            right: d[4].name,
            relationRaw: d[2],
            labelLeft,
            labelRight,
            label,
            dashed: Boolean(relationRaw.dashed),
            reversed: Boolean(relationRaw.reversed)
          } as Action
        }
      %}

classInRelation ->
    %VALID_TEXT {% (d) => ({ name: tv(d[0])}) %}

relation ->
    "<|--" {% (d) => { return { type: Relation.INHERITANCE, reversed: true } } %}
  | "<|.." {% (d) => { return { type: Relation.INHERITANCE, reversed: true, dashed: true } } %}
  | "--|>" {% (d) => { return { type: Relation.INHERITANCE } } %}
  | "..|>" {% (d) => { return { type: Relation.INHERITANCE, dashed: true } } %}
  | "*--"  {% (d) => { return { type: Relation.COMPOSITION, reversed: true } } %}
  | "*.."  {% (d) => { return { type: Relation.COMPOSITION, reversed: true, dashed: true } } %}
  | "--*"  {% (d) => { return { type: Relation.COMPOSITION } } %}
  | "..*"  {% (d) => { return { type: Relation.COMPOSITION, dashed: true } } %}
  | "o--"  {% (d) => { return { type: Relation.AGGREGATION, reversed: true } } %}
  | "o.."  {% (d) => { return { type: Relation.AGGREGATION, reversed: true, dashed: true } } %}
  | "--o"  {% (d) => { return { type: Relation.AGGREGATION } } %}
  | "..o"  {% (d) => { return { type: Relation.AGGREGATION, dashed: true } } %}
  | "-->"  {% (d) => { return { type: Relation.ASSOCIATION } } %}
  | "..>"  {% (d) => { return { type: Relation.ASSOCIATION, dashed: true } } %}
  | "<--"  {% (d) => { return { type: Relation.ASSOCIATION, reversed: true } } %}
  | "<.."  {% (d) => { return { type: Relation.ASSOCIATION, reversed: true, dashed: true } } %}
  | "--"   {% (d) => { return { type: Relation.LINK } } %}
  | ".."   {% (d) => { return { type: Relation.LINK, dashed: true } } %}

classAnnotationStatement ->
    annotation classInRelation {%
        function(d) {
          return {
            type: 'addAnnotation',
            annotation: d[0].annotation,
            className: d[1].name,
          } as Action
        }
      %}

annotation ->
    %TEXT_WITH_ANGLE_BRACKETS {%
        function(d) {
          const v = tv(d[0])
          return {
            type: 'annotation',
            annotation: v.replace(/\<\<(.*)\>\>/, '$1').trim(),
          }
        }
      %}

textInQuote ->
    %QUOTED_WORD {%
        function(d) {
          return getQuotedWord(d[0])
        }
      %}
