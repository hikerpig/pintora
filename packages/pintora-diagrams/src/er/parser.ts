import { genParserWithRules } from '../util/parser-util'
import db from './db'
import grammar from './parser/erDiagram'

export const parse = genParserWithRules(grammar, {
  dedupeAmbigousResults: true,
  postProcess(results) {
    db.apply(results)
    return results
  },
})
