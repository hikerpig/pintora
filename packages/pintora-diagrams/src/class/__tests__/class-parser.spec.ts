import { parse } from '../parser'
import db from '../db'
import { stripStartEmptyLines } from '@pintora/test-shared'

describe('class parser', () => {
  afterEach(() => {
    db.clear()
  })

  it('can parse class name', () => {
    const example = stripStartEmptyLines(`
    classDiagram
      class C1
      class N1.N2.C2
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.classes['C1']).toMatchObject({ name: 'C1' })
    expect(ir.classes['N1.N2.C2']).toMatchObject({ name: 'C2', fullName: 'N1.N2.C2', namespace: 'N1.N2' })
  })

  it('can parse methods and fields', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    class C1 {
      string field1
      int method1()

      field2: number
    }

    C1: method2()
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir).toMatchSnapshot()
    // console.log(JSON.stringify(ir, null, 2))
  })

  it('can parse member with modifier', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    class C1 {
      { static } int method1()
      {abstract} string field1
    }
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.classes['C1'].members[0].modifier).toEqual('static')
    expect(ir.classes['C1'].members[1].modifier).toEqual('abstract')
  })

  it('can parse relationship between classes', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    A1 <|-- A2
    B1 *-- B2
    C1 o-- C2
    D1 --> D2 : association
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.relations).toMatchSnapshot()
  })

  it('can parse multiple relationships', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    classA <|-- classB
    classC *-- classD
    classE o-- classF
    classG <-- classH
    classI -- classJ
    classK <.. classL
    classM <|.. classN
    classO .. classP
    class1 --|> class2
    `)
    parse(example)
  })

  it('can parse quoted labels on relations', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    C1 "1" *-- "many" C2 : contains
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.relations[0]).toMatchObject({
      type: 'addRelation',
      left: 'C1',
      right: 'C2',
      relation: 'COMPOSITION',
      labelLeft: '1',
      labelRight: 'many',
      label: 'contains',
      dashed: false,
    })
  })

  it('can parse class annotation', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    << Interface >> C1
    class C2 {
      << Serializable >>
      string test
    }
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.classes).toMatchObject({
      C1: {
        annotation: 'Interface',
      },
      C2: {
        annotation: 'Serializable',
      },
    })
  })

  it('can parse class label', () => {
    const example = stripStartEmptyLines(`
    classDiagram
    class "This is class label" as class1
    class class1 {
      RED
      Optional[Foo] foo
      void foo(a: int, b: int)
    }
    `)
    parse(example)
    const ir = db.getDiagramIR()
    expect(ir.classes).toMatchSnapshot()
  })

  it('can parse singleline note', () => {
    const backquoteExample = `classDiagram
    @note right of User: singleline note
    `
    parse(backquoteExample)
    const result = db.getDiagramIR()
    // console.log('notes', result.notes)
    expect(result.notes.length).toEqual(1)
    expect(result.notes[0]).toMatchObject({
      text: 'singleline note',
    })
  })

  it('can parse multiline note', () => {
    const multilineNoteExample = stripStartEmptyLines(`
classDiagram
  class Object {
  }
  @start_note right of Object
  aaa note -
  bbb
  @end_note
    `)
    parse(multilineNoteExample)
    const result = db.getDiagramIR()
    // console.log('notes', result.notes)
    expect(result.notes.length).toEqual(1)
    // parseMessage will trim text, so this may be somehow strange
    const messageText = result.notes[0].text
    expect(messageText).toEqual('aaa note -\nbbb')
  })

  it('can parse title', () => {
    const multilineNoteExample = stripStartEmptyLines(`
classDiagram
  title: Hello
    `)
    parse(multilineNoteExample)
    const result = db.getDiagramIR()
    expect(result.title).toBe('Hello')
  })
})
