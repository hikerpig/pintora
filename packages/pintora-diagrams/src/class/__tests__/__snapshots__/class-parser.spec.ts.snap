// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class parser can parse class label 1`] = `
{
  "class1": {
    "annotation": "",
    "fullName": "class1",
    "label": "class1",
    "members": [
      {
        "access": "public",
        "isMethod": false,
        "modifier": null,
        "name": "RED",
        "raw": "RED",
        "typeName": "",
      },
    ],
    "name": "class1",
    "namespace": "",
  },
}
`;

exports[`class parser can parse methods and fields 1`] = `
{
  "classes": {
    "C1": {
      "annotation": "",
      "fullName": "C1",
      "label": "C1",
      "members": [
        {
          "access": "public",
          "isMethod": false,
          "modifier": null,
          "name": "field1",
          "raw": "string field1",
          "typeName": "string",
        },
        {
          "access": "public",
          "isMethod": true,
          "modifier": null,
          "name": "method1()",
          "raw": "int method1()",
          "typeName": "int",
        },
        {
          "access": "public",
          "isMethod": false,
          "modifier": null,
          "name": "field2",
          "raw": "field2: number",
          "typeName": "number",
        },
        {
          "access": "public",
          "isMethod": true,
          "name": "method2()",
          "raw": "method2()",
          "typeName": "",
        },
      ],
      "name": "C1",
      "namespace": "",
    },
  },
  "configParams": [],
  "overrideConfig": {},
  "relations": [],
  "title": "",
}
`;

exports[`class parser can parse relationship between classes 1`] = `
[
  {
    "dashed": false,
    "label": "",
    "labelLeft": null,
    "labelRight": null,
    "left": "A1",
    "relation": "INHERITANCE",
    "reversed": true,
    "right": "A2",
    "type": "addRelation",
  },
  {
    "dashed": false,
    "label": "",
    "labelLeft": null,
    "labelRight": null,
    "left": "B1",
    "relation": "COMPOSITION",
    "reversed": true,
    "right": "B2",
    "type": "addRelation",
  },
  {
    "dashed": false,
    "label": "",
    "labelLeft": null,
    "labelRight": null,
    "left": "C1",
    "relation": "AGGREGATION",
    "reversed": true,
    "right": "C2",
    "type": "addRelation",
  },
  {
    "dashed": false,
    "label": "association",
    "labelLeft": null,
    "labelRight": null,
    "left": "D1",
    "relation": "ASSOCIATION",
    "reversed": false,
    "right": "D2",
    "type": "addRelation",
  },
]
`;
